### ISSUE TITLE
Double free in the PE32 parser (_free_resources)

### ISSUE THREAD

**Issue by t00sh** [radare/radare2#13035]
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Debian 4.9.130-2, AMD 64
| File format of the file you reverse (mandatory)      | PE32
| Architecture/bits of the file (mandatory)            | ARM
| r2 -v full output, **not truncated** (mandatory)     | radare2 3.3.0-git 20782 @ linux-x86-64 git.3.2.1-277-g8c022deef commit: 8c022deef84e0658394fca8746a7cbf1a13438e5 build: 2019-02-09__12:58:03

### Actual behavior

There is a potential double free in `_free_resources` in the PE32 parser.

### Steps to reproduce the behavior 

- Compile with ASAN
- Open the attached file in radare2
- Quit radare2

### Additional Logs, screenshots, source-code,  configuration dump, ...

Here is the ASAN trace : 

```
==2297==ERROR: AddressSanitizer: attempting double-free on 0x631000028800 in thread T0:
    #0 0x7f4eb91aba10 in free (/usr/lib/x86_64-linux-gnu/libasan.so.3+0xc1a10)
    #1 0x7f4eb75df570 in _free_resources /XXX/adare2-src-asan/libr/..//libr/bin/p/../format/pe/pe.c:1218
    #2 0x7f4eb23210ad in r_list_delete /XXX/radare2-src-asan/libr/util/list.c:107
    #3 0x7f4eb2321193 in r_list_purge /XXX/radare2-src-asan/libr/util/list.c:74
    #4 0x7f4eb23212f0 in r_list_free /XXX/radare2-src-asan/libr/util/list.c:83
    #5 0x7f4eb7620c7f in Pe32_r_bin_pe_free /XXX/radare2-src-asan/libr/..//libr/bin/p/../format/pe/pe.c:3526
    #6 0x7f4eb75d792f in destroy /XXX/radare2-src-asan/libr/..//libr/bin/p/bin_pe.inc:62
    #7 0x7f4eb72d8c97 in r_bin_file_free /XXX/radare2-src-asan/libr/bin/bfile.c:681
    #8 0x7f4eb23210ad in r_list_delete /XXX/radare2-src-asan/libr/util/list.c:107
    #9 0x7f4eb2321193 in r_list_purge /XXX/radare2-src-asan/libr/util/list.c:74
    #10 0x7f4eb23212f0 in r_list_free /XXX/radare2-src-asan/libr/util/list.c:83
    #11 0x7f4eb72ab3d4 in r_bin_free /XXX/radare2-src-asan/libr/bin/bin.c:550
    #12 0x7f4eb861c1ce in r_core_fini /XXX/radare2-src-asan/libr/core/core.c:2555
    #13 0x55e8cb4c314d in main /XXX/radare2-src-asan/binr/radare2/radare2.c:1559
    #14 0x7f4eb1e8d2e0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202e0)
    #15 0x55e8cb4cb1e9 in _start (/XXX/radare2-src-asan/binr/radare2/radare2+0x121e9)

0x631000028800 is located 0 bytes inside of 65536-byte region [0x631000028800,0x631000038800)
freed by thread T0 here:
    #0 0x7f4eb91aba10 in free (/usr/lib/x86_64-linux-gnu/libasan.so.3+0xc1a10)
    #1 0x7f4eb75df570 in _free_resources /XXX/radare2-src-asan/libr/..//libr/bin/p/../format/pe/pe.c:1218

previously allocated by thread T0 here:
    #0 0x7f4eb91abed0 in calloc (/usr/lib/x86_64-linux-gnu/libasan.so.3+0xc1ed0)
    #1 0x7f4eb75f3b92 in _parse_resource_directory /XXX/radare2-src-asan/libr/..//libr/bin/p/../format/pe/pe.c:2272
```
I think the problem is in the `_parse_resource_directory` function : 

```C
static void _parse_resource_directory(struct PE_(r_bin_pe_obj_t) *bin, Pe_image_resource_directory *dir, ut64 offDir, int type, int i\
d, HtUU *dirs, char *resource_name) {
  /* ... */

  for (index = 0; index < totalRes; index++) {
    /* ... */
    ut8 *resourceEntryName = NULL;
    if (entry.u1.s.NameIsString) { /* [1] */
      /* ... */
      resourceEntryName = calloc (resourceEntryNameLength + 1, sizeof (ut8));
      /* ... */
    }

    if (entry.u2.s.DataIsDirectory) { /* [2] */
      /* ... */
      if (resource_name && resourceEntryName) {
        /* ... */
        R_FREE (resource_name);
      }
      /* [3] */
      _parse_resource_directory (bin, &identEntry,
        entry.u2.s.OffsetToDirectory, type, entry.u1.Id, dirs, (char *)resourceEntryName);
      continue;
    }
    /* ... */
    r_pe_resource *rs = R_NEW0 (r_pe_resource);
    /* ... */
    if (resource_name) { /* [4] */
      rs->name = resource_name;
    } else {
      char numberbuf[6];
      rs->name = strdup (sdb_itoa (id, numberbuf, 10));
    }
    r_list_append (bin->resources, rs);
  }
}
```
Here, the `rs->name` pointer of two entries in the `bin->resources` list can reference the same address.

This is possible if a resource of type directory with the check passed in [1] contains two entries. In [4] the same pointer `resource_name` is then assigned to the two entries during the recursive call.

When the binary is freed, the two entries are passed to `_free_resources` : the 2nd trigger the double free.

Attached file: [pe32-crasher.exe.zip](https://github.com/radare/radare2/files/2847991/pe32-crasher.exe.zip)

**Comment by Maijin** [comment]
Can you send a pull-request?

**Comment by t00sh** [comment]
I tried a fix in #13037.
